### DISCUSSION TITLE
The timecode in my metadata doesnt match the time on the QR code

### DISCUSSION THREAD

**Discussion by MtnXfreeride** [gopro/labs#526] [Q&A] [answered]
I am trying to switch to DaVinci from Premiere and noticed that I have a timecode issue.  The metadata in my gopro files under "timecode first frame" and last frame do not match the time the video was made.  Somehow premiere has no issues and puts the files at the correct times on the timeline, but in DaVinci its following the timecode metadata which is far ahead of actual time. 

<img width="1920" alt="screenshot" src="https://github.com/gopro/labs/assets/144697478/4b21e28b-3e09-4cdc-ae04-798f7000dc64">

Left in the screenshot is premiere and you can see the video in/out make sense for the time shown on my phone.   The center of the screenshot is metadata from the gopro file which show a time several minutes ahead of time.  On the right is DaVinci which is reading that metadata from the file and placing clips at the wrong spots on my timeline.    Something else I noticed that was off, is that the gopro metadata shows 28.971 FPS a few lines above the timecode.   Everywhere else in the GoPro metadata is reading 29.970... this might be unreleated to my issue.

This is from a hero 11 using the latest labs firmware and before recording I showed the QR on my phone to sync it with my Hero10 which has the same results with metadata time being ahead.  

I was going towards DaVinci support first since Premiere works,  but once I saw the metadata on the file displaying the wrong time, I think the issue starts at the GoPro file level.    The big issue is that files do not align between cameras and its even putting gaps between the loop mode files on the timeline.

**Comment by dnewman-gpsw** [comment]
I agree that is wrong, but my first question is where you're getting that metadata?  What are using to extract it.  The FrameRate field is wrong, and yes, likely screwing everything up.  

This is very weird as GoPro have had timecode for 10 years now, Labs and HERO12 only adds new ways to set it.  Might be a DaVinci bug, but also odd to only be seeing this now, as I'm sure QA uses DaVinci for testing.

**Reply by MtnXfreeride** [reply]
I think I just realized this might be my issue:
https://github.com/gopro/labs/issues/483
I grabbed the metadata using the program "mediainfo".   I think the GoPro metadata is the problem since it is verifiably wrong before I even open DaVinci. Hopefully fixed in the next labs firmware, or maybe I just need to switch back to non labs firmware for a bit. ... or wait for my pair of hero 12s I ordered haha

**Reply by dnewman-gpsw** [reply]
No, this is not a Labs feature, as I pointed out timecode has been around in GoPro's for 10 years. Timecode is NOT stored in GPMF (GoPro Metadata Format), but it is in a standardized QuickTime track within the MP4.  This seems more like a DaVinci issue getting the data from the wrong place.

**Reply by MtnXfreeride** [reply]
I don't mean to push too hard on this because I have a lot of respect for how hard it is to maintain all these features.    
If I create a fresh video with my hero 11 and right click the MP4 right off the SD card before ever opening DaVinci or Premiere and launch mediainfo, I can see in the metadata that the timecode doesn't match the time the video was made.  I think DaVinci is grabbing from a DIFFRENT place than Premiere, but it literally says timecode in the metadata that matches what it is using so I think DaVinci is still correct.  Why would the metadata say "timecode" and have an irrelevant time 10 minutes in the future?  

When I mentioned timecode above I meant timecode using GoPro labs with the QR code to sync not just the metadata being written that you mention has been the case for a long time.  I didn't know if the issue I have could be specific to labs firmware because of backend stuff they built to support the QR code sync feature.

**Reply by dnewman-gpsw** [reply]
That is "mediainfo"?  

If I use FFProbe (an FFMPEG tool, the backbone of video processing everywhere), the timecode is correct.
![image](https://github.com/gopro/labs/assets/10856084/e748c139-a5f9-4cd8-92c0-5184d3b3e912)

**Reply by dnewman-gpsw** [reply]
The error is MediaInfo (I just tried it) thinking the timecode is 29 fps.  Yet when I tested with DaVinci Resolve the timecode was correct.  This was using version 18.5.1.  
![image](https://github.com/gopro/labs/assets/10856084/04467f03-a01e-4529-a5ea-6ae1cda17695)

**Reply by MtnXfreeride** [reply]
I'll do some extended testing - I installed and learned ffprobe and on the file in my example above FFProbe had the same timecode as DaVinci which was in the future.     BUT, whenever I test at home now with new files it's all correct in both FFProbe and MediaInfo.   This keeps happening when I go fishing for 3 to 4 hour blocks of time on loop mode.  I will try an extended test at home to see if I can reproduce my issue.    Thanks

**Reply by dnewman-gpsw** [reply]
If you can send, the files that fail.

**Comment by BonN2Tech** [comment]
I have a similar issue when I use QRControl to set the date/time and start recording the `1st video frame does not (most of the time) agree with FFProbe shows. Sometimes it's within a few frames and sometimes it's up to 30 before and 30 after the 1st recorded fame.  I tried using drop and TOD and see if that made any difference, but no luck.  I found that MediaInfo was way off the mark, so that's why I am using FFProbe.  see my test results here https://drive.google.com/drive/folders/13qRH8LaDCXLMxVfqFYhbFhPNsJVD9JVB?usp=drive_link

**Reply by dnewman-gpsw** [reply]
You are using HERO10, I was expect it to be +/-2 frames with reasonable consistency.  If you shutdown the camera after setting, it will be around +/- 30 frames.   HERO12 is using Dropframe, all others are TOD.

**Reply by BonN2Tech** [reply]
Thanks for that info on DF vs TOD. In my test, I jamsync first with QRControl and then begin recording. I also tried the GoPro Precision timer and it was even worse. I makes me wonder if FfProbe is returning the correct timecode. I am not sure how the metadata is stored in a string or binary and if in realtime hhmmss.ms or ?  And if Ffprobe is calculating it out. I wasn't sure of any.ofther tool I can use to inspect this

**Reply by dnewman-gpsw** [reply]
ffprobe is returning the correct timecode.  Timecode is store in frame since midnight format.

**Reply by BonN2Tech** [reply]
Thanks, David. So, then ffProbe must calculate the NDF timecode based on the frame Rate it reads from the GoPro metadata, and creates the string shown.  I noticed that in your example, there is an offset of +8 frames from the metadata time code and what is shown in the screen shot of the timer. What model was that?  Also, earlier you mentioned that to expect +/-2 frames, which I am not sure how you determined that?

**Reply by dnewman-gpsw** [reply]
It was likely an uncalibrated HERO11 or a camera that had been powered off.  I just ran a small batch of timecode scans with a HERO12 with Labs, and compared the visual timecode to the stored timecode (I used Premiere Pro, same as FFProbe, as timecode formatting is a standard Quicktime track.)

In my 8 captures the time was ahead this number of frames (at 30fps): 3,3,4,3,4,5,3,5  So an average of 3.75.  3.75/30*1000 is 125ms.  So to calibrate this camera model I set !MTCAL=125   Then my next captures had timecode offsets of: 1,0,0,0,0,0,0,0 I believe I was just lucky, but I would expect +/-1 frame precision.  The reason we don't typically worry about calibration as most users of timecode on GoPro, are syncing GoPro cameras with other GoPros. This is because no other camera supports visual timecode. A nice bridging solution is offered by Tentacle Sync with its Timebar app, which supports BLE Timecode and GoPro QR Code, they also use TCAL.

**Reply by BonN2Tech** [reply]
Those results are what I would expect.  But what /where is the  " !MTCAL" set?  Is that via QR code setting?  If all Gopro's sync together within 1 frame or so, and are consistent and the visual start frames are all the same, then would you still use that calibration?

**Reply by dnewman-gpsw** [reply]
Syncing GoPros of the same model you don't need to calibrate, they should be within 2 frames 99% of the time.  Yes, TCAL is one of the hundreds of QR code extensions.

**Reply by BonN2Tech** [reply]
[TC offsets.txt](https://github.com/gopro/labs/files/13733623/TC.offsets.txt)
Thank you. I did some more testing of my Hero 9 and 10 using both the Android app QRControl and the GoPro Github precision timer(using 30fps). They (QR code timers) both have different TC offsets as well as each camera: Hero 9 (-5 to -9 frames), whereas Hero 10 (+1 to +8) using QRControl app. If I use, instead the precisiontimer webpage, on Hero 9 (+34 to +39) and Hero 10 (+62 to + 67). Removing the bias of their respective offsets (using TCAL), the Hero 10 will be within 7 frames, and Hero 9 within 5 frames. But I am not sure what to rely on, the QRControl or the PrecisionTimer?  I've attached my teste results. As to why these cannot be as precise as the Hero12 remains to be answered. Obviously, when mixing the 2 cameras one has to perform the  time code calibration relative to each model.

**Reply by dnewman-gpsw** [reply]
There is no reason the webpage and app should be any different, they are not here, so look into your testing method.  Remember QRControl will use the time of the device it is running, same for the web page, if you are scan on one and comparing to the other all bets are off.

**Reply by BonN2Tech** [reply]
David, all videos 1st visual frame time are compared with the timecode metadata ffprobe returns for the first frame, and that's how I calculated the offset. I set/scan the QR Code on the same device (PC or phone) and compare each camera's video separately (see my test data).  So I am still puzzled why they come out different from the webpage vs the app, and why the app has a much smaller offset.

**Reply by dnewman-gpsw** [reply]
You haven't confirmed whether you're comparing the timecode stored, to the timecode displayed from the same tool that you scanned. There's a chance on a slow PC or a slow phone, the screen update is less often, but the QR code displayed will always match the timecode string on the same display, it's impossible to separate them. A timecode offset from device to another is absolutely expected.

**Reply by BonN2Tech** [reply]
Yes. I am comparing the timecode stored as metadata of the mp4 file with the timecode displayed at the 1st frame of video. You can view a sample of each QRControl and Precision Timer of the screen shots I've created. See attached
<img width="1272" alt="GH012183 using PrecisionTimer Web" src="https://github.com/gopro/labs/assets/154005982/2d044c71-e7e3-450d-9873-b5918f16c234">
<img width="1272" alt="GH012199 using QRControl" src="https://github.com/gopro/labs/assets/154005982/38ded82b-1187-4b4d-9702-c4f03716938f">

I have used two different PC's with identical results (which are reasonably fast), and certainly would not have a 2 second latency in display. I also have 500Mb/sec download speed. The QRControl runs on a Galaxy s22+ which is very fast. So, there is something else that may be causing this. Perhaps, looking at these screen shots might give some clue. If you want to inspect the original files, I can send you a link to download. Are you actually using ffProbe or relying on Premiere to decipher the starting timecode?

**Comment by dnewman-gpsw** [comment]
You're still missing, I admit, my confusing point.  The image you're using to verify the first timecode, has to be from the same source you extract the time code from. Can you confirm this?  Can you also confirm you haven't set a TCAL?

**Reply by BonN2Tech** [reply]
Yes, the image and timecode metadata shown from ffprobe are from their respective recorded files both done on the Hero 10.  Viz., GH012183.mp4 is shown with 1st frame of precisiontimer and it's corresponding metadata timecode of starting frame.  GH012199.mp4 is shown with 1st frame of QRControl timer and it's corresponding metadata timecode.  I set TCAL = 0 as I wanted to make sure there was not offset in these tests. As why precisiontimer is so far off, is more  curious to me, as I will always be using QRControl to sync my Hero 10 and Hero 9  cameras (after I adjust each using TCAL by applying the amount of offset (in ms) + or - to make them both average around 0. From my test data, the timecode of the Hero 9 is early and the Hero 10 is late.

**Reply by dnewman-gpsw** [reply, marked as answer]
Found the issue, there is a bug in QRControl which offsets a fixed 1.2 seconds to the displayed timecode, which is different to what is stored in the QR Code (causing the confusion.)  This is really old code, before there was TCAL.   I will be fixing QRControl, so calibrate with the Precision Time web page and you will do better until QRControl 2.0.1 is available.

Thank you for your patience and feedback, this is a weird bug, missed for years, and you are the first to report it.

**Reply by BonN2Tech** [reply]
I also want to thank you for your understanding and interest in helping me sort out this problem. The odd thing here, David, is my results as I've sent in the last screen shots are just the opposite with the Hero 10, when using the web Precision Timer is ahead by ~ 60 frames of the visual 1st frame  and the QRControl is only ahead by at most +8 frames of the visual 1st frame. So now I am confused!  I am curious, if you are getting the same values for Timecode as I've shown in the screenshots. Here is a link to the video used for the web Precision Timer.(https://drive.google.com/file/d/1THzDuBiGkt3otjES-w2Xz135WAQfrnCi/view?usp=sharing)

**Reply by dnewman-gpsw** [reply]
I now think that HERO10 Labs has a bug that is limiting is ability to set precision timecode.  I ported the Labs module from HERO12 back to HERO10 to see if H10 behaves better (Labs QR parser is mostly independent, from the camera version, so this step is technical.)  If you wanted to try I welcome your feedback. 

https://www.dropbox.com/scl/fi/01jzqvijbvfz97lgkf2in/HERO10LabsV15074-23Decy21.zip?rlkey=xvk4k6n3ob7ckmighvp4wtub1&dl=0

These are custom builds are unofficial, and not for general release.  Nothing has been tested, other than I saw that timecode precision was improved, and the new Labs fonts are working on H10.

**Comment by BonN2Tech** [comment]
How do I use these .bin files? Please provide more detail. Just to confirm that the Hero 9 does not accept the TCAL command (I did try, but made no difference).

**Reply by dnewman-gpsw** [reply]
Same as installing any Labs firmware, download the whole ZIP (not the individual files) and the UPDATE folder is added to the root of the SD card.

Good TCAL support limited to later modes it seems.

**Comment by BonN2Tech** [comment]
I tried twice and in both cases after seeing the gears turn for a while, a message appears "Update Failed", The Update folder is still in the SD card after it failed. I deleted it and tried again, same fail.

**Reply by dnewman-gpsw** [reply]
No idea why, I only have one HERO10 with me, and it updates fine with this UPDATE folder.  The link is only for HERO10.

**Comment by BonN2Tech** [comment]
FYI with current Labs firmware, using TCAL = -1000 setting on the web page, and resyncing with QRControl I get fairly consistent range -3 to + 1. If set to 1016 ms, should expect +/- 2 
GH012370	 0
GH012371	-3
GH012372	 0
GH012373	-3
GH012374	+1
GH012375	-1
GH012376	-2

**Reply by dnewman-gpsw** [reply]
Seems about correct.

**Comment by BonN2Tech** [comment]
Tried again to do the update. Formatted a new SD card. Recorded a clip. Then took out SD card and copied the Update folder (from your link), and same fail. I've attached the Misc folder files, if that gives any clue. Does this update take care of the 1.2sec offset or is the simply the QRcontrol issue? Since I am syncing with QRControl and using TCAL= -1000.
[qrlog.txt](https://github.com/gopro/labs/files/13759633/qrlog.txt)
[version.txt](https://github.com/gopro/labs/files/13759634/version.txt)

**Reply by dnewman-gpsw** [reply]
I guess we will have to wait for the official next release. I expect in a few months. I have no idea this install is failing.

**Comment by BonN2Tech** [comment]
I was curious if you've seen any GoPro MP4 merge programs that actually
preserve the GoPro timecode metadata in the output file? I've looked at a
few, such as mp4-merge (
https://github.com/gyroflow/mp4-merge?tab=readme-ov-file), and it seems to
work, but being written with the RUST library, I am having difficulty using
it in a C# project I've created. There were others on Github that merge,
but are not properly preserving the timecode metadata (using ffProbe to
check),   Also checking to see if there are any new builds for the Hero10
firmware that might be in beta or the issue with QRCONTROL.

**Comment by BonN2Tech** [comment]
Hello again.
I finally released my new multicam PC software that works with GoPro labs
firmware upgrade.  It will import and merge chapter files, sync and allow
you to see all views of multiple recorded videos synchronously using the
timecode from the metadata, and also imports .LRV proxies automatically. It
organizes all the clips as they are imported into projects, and you can
log, edit, and share the videos. It also has a unique feature to show an
audience what you see on multiple external displays.  You can see several
of the tutorials here
<https://youtube.com/playlist?list=PLqjGXHKP-m5ccPDuWDAe3UzrjY_k1De4X&si=yDFUQLUWkLzTTL-2>
 (
https://youtube.com/playlist?list=PLqjGXHKP-m5ccPDuWDAe3UzrjY_k1De4X&si=yDFUQLUWkLzTTL-2)
and the webpage for this product is here Multi-View HD Pro Software |
FutureVideo <https://futurevideo.tv/multi-view-hd-pro-software/>  I would
appreciate it if you could inform the GoProLabs users of this product, if
possible.  Thanks.

**Comment by BonN2Tech** [comment]
How do I get the QRControl App timecode display to show/count up to 29 frames? It only counts up to 23. I have set TOD and the frame rate button to 29.97/23.976 and Sync Now. I tried setting video to 30, but does not matter.
